### DISCUSSION TITLE
Cannot find module '@eclipse-emfcloud/sprotty-ecore/lib/action-definitions'

### DISCUSSION THREAD

**Discussion by hsd-dev** [eclipse-emfcloud/emfcloud#165] [Theia Tree Editor] [answered]
Following [this](https://github.com/eclipse-emfcloud/emfcloud/discussions/126) discussion, I am trying to get a property-view running for my GLSP diagrams based on Ecore models (some discussion here: https://github.com/eclipse-glsp/glsp/discussions/684#discussioncomment-2970972). To get the semantic URI of the selected element, I am following [EcoreGlspSelectionDataService](https://github.com/eclipse-emfcloud/ecore-glsp/blob/master/client/packages/theia-ecore/src/browser/diagram/selection-data-service.ts). As [here](https://github.com/eclipse-emfcloud/ecore-glsp/blob/master/client/packages/theia-ecore/package.json#L23), I have added `0.2.0` version of `@eclipse-emfcloud/sprotty-ecore` in my `package.json`. But I get an error
```
src/browser/diagram/theia-glsp-selection-forwarder.ts(20,8): error TS2307: Cannot find module '@eclipse-emfcloud/sprotty-ecore/lib/action-definitions' or its corresponding type declarations.
```

**Comment by ndoschek** [comment, marked as answer]
Hi @ipa-hsd!
Thank you for your interest in the `ecore-editor` and `property view`!
Currently, the `@eclipse-emfcloud/sprotty-ecore` package is [not published via npm](https://www.npmjs.com/package/@eclipse-emfcloud/sprotty-ecore).
I would suggest to implement your own `GlspSelectionDataService` in your editor application. If there arise any issues, do not hesitate to come back to this discussion!

BR, Nina

PS: Please note that the `@eclipse-emfcloud/ecore-glsp` repository contains a demonstrator for showing an editor for Ecore models. Thus, we only focus on what’s relevant for the demonstration under a controlled environment. If you are interested in transforming this demonstrator into an adoptable state, e.g. by funding the development work or working on it yourself, please let us know!

**Reply by hsd-dev** [reply]
Thank you for your response! Actually my use-case is simpler than the discussion in https://github.com/eclipse-emfcloud/emfcloud/discussions/126 I guess. I have a jar file from the eclipse based EMF project where I have the ecore models defined and implemented the GLSP server. I want to get the JSON forms property view for the selected elements in the graphical editor. Since I am not using emfcloud and hence don't have ModelServer in my project, which parts of the eclipse-glsp project do I need to re-implement?

Is your response [here](https://github.com/eclipse-emfcloud/emfcloud/discussions/126#discussioncomment-2308108) still valid?

**Reply by ndoschek** [reply]
I'm not sure if I can follow. Do you want to use the property-view as a standalone widget in the Theia editor?

> Is your response [here](https://github.com/eclipse-emfcloud/emfcloud/discussions/126#discussioncomment-2308108) still valid?

If so, yes, this is the go-to way to implement the property-view as part of your Theia editor, that also supports a GLSP diagram editor.

The `GlspSelectionDataService` makes sure that the selected element in the GLSP diagram editor is forwarded to the global Theia selection. In the case of the Ecore editor, the `GlspSelectionDataService` resolves via the `RequestSemanticUriAction` the graphic element id of the selected element and returns the `semanticUri` of the semantic element. In that case this is the xmi id of the Ecore element, see also the corresponding [`RequestSemanticUriActionHandler`](https://github.com/eclipse-emfcloud/ecore-glsp/blob/master/server/org.eclipse.emfcloud.ecore.glsp/src/main/java/org/eclipse/emfcloud/ecore/glsp/handler/RequestSemanticUriActionHandler.java).

If the selection data is correctly forwarded by the previously mentioned `GlspSelectionDataService`, the `PropertyDataService` is responsible to retrieve the data based on the provided `selection` object, which in the case of the [`EcoreGlspPropertyDataService`](https://github.com/eclipse-emfcloud/ecore-glsp/blob/master/client/packages/ecore-property-view/src/browser/property-data-service.ts) is the `semanticUri`. Based on this `semanticUri`, the data is retrieved via the ModelServer client API
If you do not use the ModelServer, you have to find another service that fetches and provides the property data for the selected element.

I would also recommend to take a look at the [Theia documentation](https://theia-ide.org/docs/property_view) about the concept of the Theia property view and the [`GLSP documentation`](https://www.eclipse.org/glsp/documentation/actionhandler/) about actions and action handlers.

BR, Nina

**Reply by hsd-dev** [reply]
OK I was able to implement the `GlspSelectionDataService` based on the existing example and also `RequestSemanticUriActionHandler`
```java
public class RequestSemanticUriActionHandler extends AbstractActionHandler<RequestSemanticUriAction> {

    @Inject
    protected EMFNotationModelState modelState;

    @Override
    protected List<Action> executeAction(RequestSemanticUriAction action) {
        List<Action> actionList = new ArrayList<>();

        String modelUri = EcoreUtil.getURI(modelState.getSemanticModel()).toFileString();

        EMFNotationModelIndex modelIndex = modelState.getIndex();
        Optional<GModelElement> element = modelIndex.findElementByClass(action.getGraphicElementId(),
            GModelElement.class);
        if (element.isPresent()) {
            Optional<EObject> semanticElement = modelIndex.getEObject(element.get());

            String fragment = semanticElement.get().eResource().getURIFragment(semanticElement.get());
            actionList.add(
                new SetSemanticUriAction(modelUri, fragment, semanticElement.get().eClass().getName()));
        }
        return actionList.isEmpty() ? none() : actionList;
    }

}
```
Correct me if I am wrong. Also, I think I misunderstood what `ModelServer` is. This documentation was helpful: https://eclipsesource.com/blogs/2021/02/25/the-emf-cloud-model-server/. In my case I do have text and graphical editors which will be working on the same model instance, so I think I need a model server as well. In that case, will the model server replace the existing GLSP and Xtext servers? 

Edit: This sentence in the documentation says
"The diagram editor is based on [Eclipse GLSP](https://www.eclipse.org/glsp/). In this case, a Java-based server is subscribed to the model server that itself updates the browser client."
So basically, I have to "subscribe" the GLSP server to the model server somehow. Right?

**Reply by ndoschek** [reply]
If you want to support multiple editor capabilities (as you mentioned diagram editing, text editing, perhaps tree editing) then the Model Server is a good choice as managing component of the model instances (see also the info graphic in [the linked blog post](https://eclipsesource.com/blogs/2021/02/25/the-emf-cloud-model-server/)).
This will not replace the GLSP server, but the GLSP server will retrieve its source models from the Model Server.
And yes, the GLSP server can and should also subscribe to the Model Server to be notified about any updates that may be caused by other components (e.g. the textual editor).

PS: We are currently working on a simple project template that combines the essentials for this use case: model management by the Model Server and integration of the GLSP server and should be available in the course of August.

**Reply by hsd-dev** [reply]
Alright, looks like I am on the right track then. Since `ecore-glsp` is based on GLSP, it already implements this connection between GLSP server and model server right? Could you please point to the relevant code?

**Reply by hsd-dev** [reply]
I have another noob question: Why is `SetSemanticUriAction` necessary? Since we already get the type of the element selected
```java
semanticElement.get().eClass()
```
can't we return the properties of the class as a response action? In case there is a single editor (e.g. GLSP), model server would be an overkill just to get the properties, isn't?

> If you do not use the ModelServer, you have to find another service that fetches and provides the property data for the selected element.

I think for now, I would prefer to implement another service with request-response to get the properties. Would be better to wait to implement model server till you have the project template ready.

**Reply by ndoschek** [reply]
As I mentioned earlier, it depends on which kinds of editor capabilities you want to support.
Of course, if you plan to support a diagram editor only, you could omit the Model Server and let the GLSP server handle the model management (and also property retrieval for example).

> Why is `SetSemanticUriAction` necessary?

The set action is the response action in this case and is necessary to fulfill the request-response pattern of GLSP actions, please have a look at the documentation of GLSP actions: https://www.eclipse.org/glsp/documentation/actionhandler/.

In the meantime, to support an EMF model via a GLSP server you could have a look at the current project templates, described here: [Eclipse GLSP Getting Started](https://www.eclipse.org/glsp/documentation/gettingstarted/)
The formerly mentioned template with the integration of the Model Server will be listed here once it is ready as well.

**Reply by hsd-dev** [reply]
[This](https://github.com/eclipse-emfcloud/emfcloud-modelserver/blob/bf024daedff46b0a3917e0d0f86b459a386ca21b/bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/DefaultModelController.java#L175-L187) is where the element is returned as JSON in the model server
```Java
   @Override
   public void getModelElementById(final Context ctx, final String modeluri, final String elementid) {
      Optional<EObject> element = this.modelRepository.getModelElementById(modeluri, elementid);
      if (element.isEmpty()) {
         notFound(ctx, "Element with id '" + elementid + "' of model '" + uriConverter.deresolveModelURI(ctx, modeluri)
            + "' not found!");
         return;
      }
      try {
         success(ctx, codecs.encode(modeluri, ctx, element.get()));
      } catch (EncodingException exception) {
         encodingError(ctx, exception);
      }
   }
```
Based on that I wrote an action handler in my GLSP server which receives action request from the `providePropertyData` in `PropertyDataService` defined on the client side:
```js
...
@inject(TYPES.IActionDispatcher) protected actionDispatcher: GLSPActionDispatcher;
...
    async providePropertyData(selection: Record<string, unknown> | undefined): Promise<Record<string, unknown> | undefined> {
        if (selection && isGlspSelection(selection) && selection.selectedElementsIDs) {
            const selectionData = this.getSelectionData(selection);
            return new Promise(resolve => {
                 this.actionDispatcher.request(new RequestModelElementAction(selectionData.semanticUri));
            });
        }
        return undefined;
    }
```
But when I click on the graphical element in the editor, I get an error
```
root ERROR Error: No matching bindings found for serviceIdentifier: Symbol(IActionDispatcher)
    at _validateActiveBindingCount (http://localhost:3000/bundle.js:78687:23)
    at _getActiveBindings (http://localhost:3000/bundle.js:78673:5)
    at _createSubRequests (http://localhost:3000/bundle.js:78716:26)
    at http://localhost:3000/bundle.js:78740:17
    at Array.forEach (<anonymous>)
    at http://localhost:3000/bundle.js:78739:26
    at Array.forEach (<anonymous>)
    at _createSubRequests (http://localhost:3000/bundle.js:78719:20)
    at Object.plan (http://localhost:3000/bundle.js:78761:9)
    at http://localhost:3000/bundle.js:78283:37
```

**Reply by sgraband** [reply]
Hey @ipa-hsd,

where is the code in which you inject the `Types.IActionDispatcher` located? On the GLSP Server? Please note, that the `Types.IActionDispatcher` is used for sending actions from the client to the server and hence cannot be used on the Server. Please see here for reference => https://www.eclipse.org/glsp/documentation/actionhandler/#overview.

If the code is on the client, it would be nice to know your bindings, to see why it seems to be unbound. Hope this helps.

**Reply by hsd-dev** [reply]
I have created a package similar to [ecore-property-view](https://github.com/eclipse-emfcloud/ecore-glsp/tree/master/client/packages/ecore-property-view) (so on the client side) and instead of using the [modelServerClient](https://github.com/eclipse-emfcloud/ecore-glsp/blob/master/client/packages/ecore-property-view/src/browser/property-data-service.ts#L50), I replaced it with action as shown above. And in my frontend module in that package, I bind the property data service and the widget provider (which at the moment doesn't do much)
```js
export default new ContainerModule(bind => {
    bind(PropertyDataService).to(EcoreGlspPropertyDataService).inSingletonScope();
    bind(PropertyViewWidgetProvider).to(EcoreGlspPropertyViewWidgetProvider).inSingletonScope();
});
```
I have implemented [EcoreGlspSelectionDataService](https://github.com/eclipse-emfcloud/ecore-glsp/blob/master/client/packages/theia-ecore/src/browser/diagram/selection-data-service.ts) with `RequestSemanticUriAction` in my `minimal-theia` package which works fine (it returns me `semanticUri`).

So my package structure is:
```
ecore-property-view
minimal-glsp
minimal-theia (which implements the GLSP selection service)
```
Please note that even though the package names have "minimal", I have my own ecore model defined in the GLSP server.

**Reply by hsd-dev** [reply]
Can I inject the dispatcher in multiple packages?
```js
@inject(TYPES.IActionDispatcher) protected actionDispatcher: GLSPActionDispatcher;
```

**Reply by sgraband** [reply]
Just so we are on the same page: You are trying to fill yout your property view, by dispatching a GLSP action from the property view?

This will not work with your current approach, as the container of the `ecore-property-view` does not have any GLSP bindings (as you only bind the `PropertyDataService` and the `PropertyViewWidgetProvider` in there.

This was never really a use case for us, as this is a prime example for using the model-server, like Nina mentioned before. The GLSP editor and the properties view would both connect to the model server and get the data from there.

If you don't want to go that route you would need to get the relevant GLSP bindings, which are distinct between different diagram widgets, so you need to retrieve the current widget you want to use and get the DI container from that widget (which should be exposed by the widget).

**Reply by hsd-dev** [reply]
Thank you, that answers my question. Even though I am curious on how to get the "relevant GLSP bindings", implementing the model server is definitely a long term solution since I intend to synchronize my existing text editor with the graphical editor in the future.

**Reply by sgraband** [reply]
For your use case the `ActionDispatcher` of the widget should be enough. You can easily get this from the DiagramWidget https://github.com/eclipse/sprotty-theia/blob/efd153a410267d81af555e70c219e9efa3e3b197/src/theia/diagram-widget.ts#L68. Your GLSP widget likely either extends `DiagramWidget` or `GLSPDiagramWidget` so that method should be accessible once you get the widget.

The widget you can get with something similar to [this](https://github.com/eclipse-theia/theia/discussions/8409).

Hope this helps

**Reply by hsd-dev** [reply]
wow thank you so much @sgraband and @ndoschek! This worked for me:
```js
const widget = this.shell.currentWidget as DiagramWidget;
widget.actionDispatcher.request(new RequestModelElementAction(selectionData.semanticUri));
```
Have a nice weekend! <3

**Reply by hsd-dev** [reply]
@sgraband sorry to reopen this issue, but I am stuck at this problem. The action handlers are working well now, but the property view is not rendered properly. The response of the `RequestModelElementAction` is a string
```json
{"name":"NewLink0","id":"95e0cf3b-5e34-4f5d-9a4d-b6a21c6b9af0"}
```
Please note that the function signature of `providePropertyData` is
```js
async providePropertyData(selection: string | undefined): Promise<string | undefined>
```
In [this](https://github.com/eclipse-emfcloud/emfcloud/discussions/126#discussioncomment-2308108) discussion is it mentioned: 
"But as JsonForms is capable of generating the type schema and UI schema itself, based on the given JSON object (in our case the property object). So to start off, we can return undefined and let JsonForms generate the schemas temporarily."

So at the moment I return `undefined` from `getSchema` and `getUiSchema` functions. And my widget provider is basic
```js
@injectable()
export class EcoreGlspPropertyViewWidgetProvider extends JsonFormsPropertyViewWidgetProvider {

    canHandle(selection: Record<string, unknown> | undefined): number {
        return isGlspSelection(selection) ? 100 : 0;
    }

}
```
 But the view that is rendered looks like this
![Screenshot from 2022-08-11 10-52-21](https://user-images.githubusercontent.com/31062848/184100313-e1068d92-7110-4648-b6a1-a22768da4141.png)

**Reply by sgraband** [reply]
Hey @ipa-hsd,

JsonForms expects a `JSONObject` and not a string as its data. You could change this by using for example `JSON.parse(<your string>)` . When you pass the Object instead of the string to JsonForms it should work. Hope this helps.

**Reply by hsd-dev** [reply]
I thought so, but I was facing [this](https://github.com/eclipse-emfcloud/emfcloud/discussions/125) problem. I changed it to `any` and now it works. Thanks for saving the day!
